chromium/background.js

```diff
@@ -618,3 +647,97 @@ chrome.runtime.onConnect.addListener(function (port) {
     });
   }
 });
+
+// This is necessary for communication with the popup in Firefox Private
+// Browsing Mode, see https://bugzilla.mozilla.org/show_bug.cgi?id=1329304
+chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
+  if (message.type == "get_option") {
+    storage.get(message.object, function(item){
+      sendResponse(item);
+    });
+    return true;
+  } else if (message.type == "set_option") {
+    storage.set(message.object);
+  } else if (message.type == "get_is_extension_enabled") {
+    sendResponse(isExtensionEnabled);
+  } else if (message.type == "set_is_extension_enabled") {
+    isExtensionEnabled = message.object;
+    sendResponse(isExtensionEnabled);
+  } else if (message.type == "delete_from_ruleset_cache") {
+    all_rules.ruleCache.delete(message.object);
+  } else if (message.type == "get_active_rulesets_and_hostnames") {
+    sendResponse({
+      rulesets: activeRulesets.getRulesets(message.object),
+      hostnames: activeRulesets.getHostnames(message.object)
+    });
+  } else if (message.type == "set_ruleset_active_status") {
+    var ruleset = activeRulesets.getRulesets(message.object.tab_id)[message.object.name];
+    ruleset.active = message.object.active;
+    sendResponse(true);
+  } else if (message.type == "add_new_rule") {
+    addNewRule(message.object, function() {
+      sendResponse(true);
+    });
+    return true;
+  } else if (message.type == "update_state") {
+    updateState();
+  } else if (message.type == "remove_rule") {
+    removeRule(message.object);
+  } else if (message.type == "import_settings") {
+    import_settings(message.object).then(() => {
+      sendResponse(true);
+    });
+  }
+});
+
+// Send a message to the embedded webextension bootstrap.js to get settings to import
+chrome.runtime.sendMessage("import-legacy-data", function(settings){
+  import_settings(settings);
+});
+
+/**
+ * Enable switch planner for specific tab
+ * @param settings the settings object
+ */
+async function import_settings(settings){
+  if(settings.changed){
+    // Load custom rulesets and add to storage
+    await new Promise(resolve => {
+      for(let ruleset of settings.custom_rulesets){
+        all_rules.addFromXml((new DOMParser()).parseFromString(ruleset, 'text/xml'));
+      }
+      storage.set({"legacy_custom_rulesets": settings.custom_rulesets}, item => {
```

Comment:
@jsha the other custom rulesets are specially formatted objects that do not allow specifying, for instance, exclusions.  Here is one example:

```js
[{"host":"example.com","redirectTo":"https://example.com/","urlMatcher":"^http://example\\.com/"}]
```

This was never in XML format, since it is added via the Chrome UI.  From that format, a `RuleSet` object is created.  This format, notably, does not in its current form support exclusions & securecookies.  We'd have to build support for that, as well as code that converts XML to the `custom_rule` format.  I'm not against doing this, but it seems like the job for a future migration.